# ch03_PyTorch进阶教程

## 3.1 Broadcasting 广播 

In [41]:
import torch
import numpy

In [5]:
a = torch.ones(2,3)
print(a)
b = a+5
print(b)

tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[6., 6., 6.],
        [6., 6., 6.]])


### Note1: Broadcasting
- situation 1:
 - tensorA : [4,32,14,14]
 - tensorB : [1,32,1,1] -> [4,32,14,14]
- situation 2:
 - tensorA : [4,32,14,14]
 - tensorB : [14,14] -> [1,1,14,14] -> [4,32,14,14]
- situation 3:
 - tenosrA : [4,32,14,14]
 - tensorB : [2,32,14,14] -> 手动完成 + B[0]

## 3.2: 拼接与拆分
- cat
- stack
- split
- chunk

In [30]:
# cat
print('cat函数,不创建新的维度:')
a1 = torch.rand(4,3,32,32)
a2 = torch.rand(4,3,32,32)
print('a1: ',a1.shape,'\na2: ',a2.shape)
print('torch.cat([a1,a2],dim=0): ',torch.cat([a1,a2],dim=0).shape,'\n')

print('stack函数,会创建新的维度:')
print('a1: ',a1.shape,'\na2: ',a2.shape)
print('torch.stack([a1,a2],dim=2): ',torch.stack([a1,a2],dim=0).shape,'\n')

print('splic函数,根据长度拆分:')
c = torch.rand(3,32,8)
print('c: ',c.shape)
c1,c2 = c.split([2,1],dim=0)
print('c1,c2 = c.split([2,1],dim=0): ','c1:',c1.shape,'  c2:',c2.shape)
print('上面这种写法,给split传入一个列表,[2,1]表示在第0维分成两个tenor,一个长为2一个长为1\n')


d = torch.rand(6,32,8)
print('d: ',d.shape)
d1,d2 = d.split(3,dim=0)
print('d1,d2 = d.split(3,dim=0): ','d1:',d1.shape,'  d2:',d2.shape)
print('上面这种写法,3表示长度,在第0维将d分成6/3=2个tensor,每个tensor长为3\n')

print('chunk函数,按数量拆分:')
a = torch.rand(8,32,8)
a1,a2 = a.chunk(2,dim=0)
print('a1: ',a1.shape,'\na2: ',a2.shape)
print('a1,a2=a.chunk(2,dim=0): ','a1: ',a1.shape,'  a2:',a2.shape)

cat函数,不创建新的维度:
a1:  torch.Size([4, 3, 32, 32]) 
a2:  torch.Size([4, 3, 32, 32])
torch.cat([a1,a2],dim=0):  torch.Size([8, 3, 32, 32]) 

stack函数,会创建新的维度:
a1:  torch.Size([4, 3, 32, 32]) 
a2:  torch.Size([4, 3, 32, 32])
torch.stack([a1,a2],dim=2):  torch.Size([2, 4, 3, 32, 32]) 

splic函数,根据长度拆分:
c:  torch.Size([3, 32, 8])
c1,c2 = c.split([2,1],dim=0):  c1: torch.Size([2, 32, 8])   c2: torch.Size([1, 32, 8])
上面这种写法,给split传入一个列表,[2,1]表示在第0维分成两个tenor,一个长为2一个长为1

d:  torch.Size([6, 32, 8])
d1,d2 = d.split(3,dim=0):  d1: torch.Size([3, 32, 8])   d2: torch.Size([3, 32, 8])
上面这种写法,3表示长度,在第0维将d分成6/3=2个tensor,每个tensor长为3

chunk函数,按数量拆分:
a1:  torch.Size([4, 32, 8]) 
a2:  torch.Size([4, 32, 8])
a1,a2=a.chunk(2,dim=0):  a1:  torch.Size([4, 32, 8])   a2: torch.Size([4, 32, 8])


## 3.3 数学运算
- add/sub/mul/div   +-*/
- matmul 矩阵相乘
 - torch.mm (only for 2d)
 - torch.matmul (推荐)
 - @ (numpy)
- pow sqrt/rsqrt
- exp log log2 log10
- round
 - .floor()  .ceil()
 - .round()
 - .trunc()  .frac()

In [54]:
print('matmul矩阵乘法:')
a = torch.rand(4,3,28,64)
b = torch.rand(4,3,64,32)
print('a.shape: ',a.shape,'  b.shape: ',b.shape)
print('对于2D以上的tensor乘法,只能用matmul,不能用mm,matmul只取最后两维:')
print('torch.matmul(a,b) 即[28,64]*[64,32]: ',torch.matmul(a,b).shape,'\n')

matmul矩阵乘法:
a.shape:  torch.Size([4, 3, 28, 64])   b.shape:  torch.Size([4, 3, 64, 32])
对于2D以上的tensor乘法,只能用matmul,不能用mm,matmul只取最后两维:
torch.matmul(a,b) 即[28,64]*[64,32]:  torch.Size([4, 3, 28, 32]) 



In [55]:
print('power用法:')
a = torch.full([2,2],16)
print('a:\n',a)
print('a**2:\n',a**2)
print('a.pow(2):\n',a.pow(2))
print('a.sqrt():\n',a.sqrt())
print('a.rsqrt() 平方根的倒数:\n',a.rsqrt())

power用法:
a:
 tensor([[16., 16.],
        [16., 16.]])
a**2:
 tensor([[256., 256.],
        [256., 256.]])
a.pow(2):
 tensor([[256., 256.],
        [256., 256.]])
a.sqrt():
 tensor([[4., 4.],
        [4., 4.]])
a.rsqrt() 平方根的倒数:
 tensor([[0.2500, 0.2500],
        [0.2500, 0.2500]])


In [56]:
print('exp log 用法:')
a = torch.exp(torch.ones(2,2))
print('a = torch.exp(torch.ones(2,2)):\n',a)
print('torch.log(a):\n',torch.log(a))
print('torch.log10(a):\n',torch.log10(a))
print('torch.log2(a):\n',torch.log2(a))

exp log 用法:
a = torch.exp(torch.ones(2,2)):
 tensor([[2.7183, 2.7183],
        [2.7183, 2.7183]])
torch.log(a):
 tensor([[1.0000, 1.0000],
        [1.0000, 1.0000]])
torch.log10(a):
 tensor([[0.4343, 0.4343],
        [0.4343, 0.4343]])
torch.log2(a):
 tensor([[1.4427, 1.4427],
        [1.4427, 1.4427]])


In [67]:
print('floor() 和 ceil() 用法:')
a = torch.tensor(3.14)
print('a: ',a)
print('a.floor()向上取整: ',a.floor())
print('a.ceil()向下取整: ',a.ceil())

print('\n')

print('trunc和frac将小数裁剪,trunc取整数部分,frac取小数部分:')
print('a.trunc(): ',a.trunc())
print('a.frac(): ',a.frac())

print('\n')

print('round四舍五入:')
print('a.round(): ',a.round())

print('\n')

print('clamp多用于梯度裁剪:')
grad = torch.rand(2,3)*15
print('grad: \n',grad)
print('grad.max(): ',grad.max())
print('grad.median() 中间值: ',grad.median(),'\n')
print('grad.clamp(10),小于10的都赋值为10: \n',grad.clamp(10),'\n')
print('grad.clamp(5,10),小于5的赋值为5,大于是10的赋值为10: \n',grad.clamp(5,10),'\n')

floor() 和 ceil() 用法:
a:  tensor(3.1400)
a.floor()向上取整:  tensor(3.)
a.ceil()向下取整:  tensor(4.)


trunc和frac将小数裁剪,trunc取整数部分,frac取小数部分:
a.trunc():  tensor(3.)
a.frac():  tensor(0.1400)


round四舍五入:
a.round():  tensor(3.)


clamp多用于梯度裁剪:
grad: 
 tensor([[ 5.7228,  0.5189,  1.9334],
        [14.1330,  3.7156, 11.3122]])
grad.max():  tensor(14.1330)
grad.median() 中间值:  tensor(3.7156) 

grad.clamp(10),小于10的都赋值为10: 
 tensor([[10.0000, 10.0000, 10.0000],
        [14.1330, 10.0000, 11.3122]]) 

grad.clamp(5,10),小于5的赋值为5,大于是10的赋值为10: 
 tensor([[ 5.7228,  5.0000,  5.0000],
        [10.0000,  5.0000, 10.0000]]) 



## 3.4 属性统计
- norm
- mean sum
- prod
- max,min,argmin,argmax
- kthvalue,topk

In [77]:
print('norm 范数:')
a = torch.full([8],1)
b = a.view(2,4)
c = a.view(2,2,2)
print('a:\n',a)
print('a的1范数为所有元素的绝对值之和 a.norm(1): ',a.norm(1))
print('a的2范数为所有元素绝对值的平方和开根号 a.norm(2): ',a.norm(2))

print('\nb:\n',b)
print('b.norm(1,dim=1) b在第一维上的1范数: ',b.norm(1,dim=1))

print('\nc:',c)
print('c.norm(2,dim=0) c在第0维上的2范数: ',c.norm(2,dim=0))

norm 范数:
a:
 tensor([1., 1., 1., 1., 1., 1., 1., 1.])
a的1范数为所有元素的绝对值之和 a.norm(1):  tensor(8.)
a的2范数为所有元素绝对值的平方和开根号 a.norm(2):  tensor(2.8284)

b:
 tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.]])
b.norm(1,dim=1) b在第一维上的1范数:  tensor([4., 4.])

c: tensor([[[1., 1.],
         [1., 1.]],

        [[1., 1.],
         [1., 1.]]])
c.norm(2,dim=0) c在第0维上的2范数:  tensor([[1.4142, 1.4142],
        [1.4142, 1.4142]])


In [100]:
a = torch.arange(8).view(2,4).float()
print('a: ',a)
print('a.min(): ',a.min())
print('a.max(): ',a.max())
print('a.sum(): ',a.sum())
print('a.mean(): ',a.mean())
print('a.prod()累乘: ',a.prod())
print('a.argmax() flatten后最大值的索引: ',a.argmax())
print('a.argmin() flatten后最小值的索引: ',a.argmin())
print('a.argmax(dim=1) 第一维度: ',a.argmax(dim=0))

print('\n')

a = torch.rand(4,10)
print('a:\n',a,'\n')
print('a.max(dim=1): \n',a.max(dim=1),'\n')
print('a.argmax(dim=1: )',a.argmax(dim=1),'\n')

print('a.max(dim=1,keepdim=True): \n',a.max(dim=1,keepdim=True),'\n')
print('a.argmax(dim=1,keepdim=True): ',a.argmax(dim=1,keepdim=True))

a:  tensor([[0., 1., 2., 3.],
        [4., 5., 6., 7.]])
a.min():  tensor(0.)
a.max():  tensor(7.)
a.sum():  tensor(28.)
a.mean():  tensor(3.5000)
a.prod()累乘:  tensor(0.)
a.argmax() flatten后最大值的索引:  tensor(7)
a.argmin() flatten后最小值的索引:  tensor(0)
a.argmax(dim=1) 第一维度:  tensor([1, 1, 1, 1])


a:
 tensor([[0.2580, 0.2123, 0.3272, 0.8038, 0.2387, 0.2304, 0.3114, 0.7600, 0.7949,
         0.4442],
        [0.9829, 0.9999, 0.4356, 0.5968, 0.3660, 0.2467, 0.7214, 0.9781, 0.3954,
         0.3355],
        [0.3723, 0.3252, 0.3333, 0.2286, 0.2606, 0.6997, 0.1471, 0.9226, 0.4864,
         0.8233],
        [0.1116, 0.4671, 0.5081, 0.5448, 0.2634, 0.3047, 0.4667, 0.9692, 0.4291,
         0.0211]]) 

a.max(dim=1): 
 torch.return_types.max(
values=tensor([0.8038, 0.9999, 0.9226, 0.9692]),
indices=tensor([3, 1, 7, 7])) 

a.argmax(dim=1: ) tensor([3, 1, 7, 7]) 

a.max(dim=1,keepdim=True): 
 torch.return_types.max(
values=tensor([[0.8038],
        [0.9999],
        [0.9226],
        [0.9692]]),
indices=

In [106]:
print('topk or kthvalue用法:')
a = torch.rand(2,6)
print('a:\n',a,'\n')
print('a.topk(3,dim=1)在第一维上取概率前3大的元素:\n',a.topk(3,dim=1),'\n')

print('a.topk(3,dim=1,largest=False)在第一维取最小的3个元素:\n',a.topk(3,dim=1,largest=False),'\n')

print('a.kthvalue(6,dim=1)在第一维上取第6小的元素,即第1大,默认是从小到大排序:\n',
     a.kthvalue(6,dim=1),'\n')

print('a.kthvalue(6,dim=1,keepdim=True)在第一维上取第6小的元素,即第1大,默认是从小到大排序:\n',
     a.kthvalue(6,dim=1,keepdim=True),'\n')

topk or kthvalue用法:
a:
 tensor([[0.6375, 0.0859, 0.2007, 0.9008, 0.3690, 0.4261],
        [0.0794, 0.5686, 0.8968, 0.2543, 0.5664, 0.6118]]) 

a.topk(3,dim=1)在第一维上取概率前3大的元素:
 torch.return_types.topk(
values=tensor([[0.9008, 0.6375, 0.4261],
        [0.8968, 0.6118, 0.5686]]),
indices=tensor([[3, 0, 5],
        [2, 5, 1]])) 

a.topk(3,dim=1,largest=False)在第一维取最小的3个元素:
 torch.return_types.topk(
values=tensor([[0.0859, 0.2007, 0.3690],
        [0.0794, 0.2543, 0.5664]]),
indices=tensor([[1, 2, 4],
        [0, 3, 4]])) 

a.kthvalue(6,dim=1)在第一维上取第6小的元素,即第1大,默认是从小到大排序:
 torch.return_types.kthvalue(
values=tensor([0.9008, 0.8968]),
indices=tensor([3, 2])) 

a.kthvalue(6,dim=1,keepdim=True)在第一维上取第6小的元素,即第1大,默认是从小到大排序:
 torch.return_types.kthvalue(
values=tensor([[0.9008],
        [0.8968]]),
indices=tensor([[3],
        [2]])) 



In [111]:
print('比较函数:')
a = torch.rand(3,3)
print('a:\n',a)

print('a>0:\n',a>0)
print('torch.gt(a,0):\n',torch.gt(a,0))
print('\n\n')

a = torch.ones(2,3)
b = torch.rand_like(a)
print('a:\n',a)
print('b:\n',b)
print('torch.eq(a,b): ',torch.eq(a,b))
print('torch.eq(a,a): ',torch.eq(a,a))
print('torch.equal(a,a): ',torch.equal(a,a))

比较函数:
a:
 tensor([[0.2238, 0.6011, 0.9244],
        [0.0446, 0.0427, 0.3566],
        [0.9677, 0.6089, 0.1657]])
a>0:
 tensor([[True, True, True],
        [True, True, True],
        [True, True, True]])
torch.gt(a,0):
 tensor([[True, True, True],
        [True, True, True],
        [True, True, True]])



a:
 tensor([[1., 1., 1.],
        [1., 1., 1.]])
b:
 tensor([[0.4793, 0.5230, 0.0222],
        [0.3620, 0.8653, 0.4292]])
torch.eq(a,b):  tensor([[False, False, False],
        [False, False, False]])
torch.eq(a,a):  tensor([[True, True, True],
        [True, True, True]])
torch.equal(a,a):  True


## 3.5 高阶操作
- where
- gather

In [116]:
print('where : torch.where(condition,x,y) -> Tensor \n')

cond = torch.tensor([[0.6,0.7],[0.8,0.4]])
a = torch.zeros(2,2)
b = torch.ones(2,2)*100
print('cond:\n',cond)
print('a:\n',a)
print('b:\n',b)
print('\n torch.where(cond>0.5,a,b):\n',torch.where(cond>0.5,a,b))

where : torch.where(condition,x,y) -> Tensor 

cond:
 tensor([[0.6000, 0.7000],
        [0.8000, 0.4000]])
a:
 tensor([[0., 0.],
        [0., 0.]])
b:
 tensor([[100., 100.],
        [100., 100.]])

 torch.where(cond>0.5,a,b):
 tensor([[  0.,   0.],
        [  0., 100.]])


In [128]:
print('gather : torch.gather(input,dim,index,out=None) -> Tensor')
prob = torch.rand(4,10)
idx = prob.topk(dim=1,k=3)
print('prob:\n',prob,'\n')
print('idx:\n',idx,'\n')
idx = idx[1]
print('idx=idx[1]:\n',idx,'\n')

label = torch.arange(10)+100
print('label:\n',label,'\n')
label = label.expand(4,10)
print('label.expand(4,10):\n',label,'\n')

g = torch.gather(label,dim=1,index=idx.long())
print('torch.gather(label.expand(4,10),dim=1,index=idx.long()):\n',g)

gather : torch.gather(input,dim,index,out=None) -> Tensor
prob:
 tensor([[0.5026, 0.8827, 0.5593, 0.0066, 0.9751, 0.6705, 0.3195, 0.8601, 0.5605,
         0.1815],
        [0.7326, 0.2979, 0.8574, 0.7390, 0.2142, 0.2609, 0.8189, 0.1418, 0.9417,
         0.9883],
        [0.5461, 0.6960, 0.3400, 0.4376, 0.1514, 0.4703, 0.3769, 0.6260, 0.9581,
         0.5978],
        [0.4863, 0.7357, 0.8359, 0.5174, 0.3940, 0.6628, 0.4596, 0.8769, 0.4706,
         0.0394]]) 

idx:
 torch.return_types.topk(
values=tensor([[0.9751, 0.8827, 0.8601],
        [0.9883, 0.9417, 0.8574],
        [0.9581, 0.6960, 0.6260],
        [0.8769, 0.8359, 0.7357]]),
indices=tensor([[4, 1, 7],
        [9, 8, 2],
        [8, 1, 7],
        [7, 2, 1]])) 

idx=idx[1]:
 tensor([[4, 1, 7],
        [9, 8, 2],
        [8, 1, 7],
        [7, 2, 1]]) 

label:
 tensor([100, 101, 102, 103, 104, 105, 106, 107, 108, 109]) 

label.expand(4,10):
 tensor([[100, 101, 102, 103, 104, 105, 106, 107, 108, 109],
        [100, 101, 102, 103, 1

### Note:
gather查表,idx中第一行为[4,1,7],在label中,第四个元素为104,第一个元素为101,第七个元素是107,所以gather得到的第一行是[104,101,107],以此类推